src/sorted_iterator.rs

```diff
@@ -173,13 +173,16 @@ where
     fn size_hint(&self) -> (usize, Option<usize>) {
         self.bh.iter().fold((0, Some(0)), |(lo, hi), it| {
             let (ilo, ihi) = it.size_hint();
-            (max(lo, ilo), hi.and_then(|hi| ihi.map(|ihi| hi + ihi)))
+            (
+                max(lo, ilo),
+                hi.and_then(|hi| ihi.and_then(|ihi| hi.checked_add(ihi))),
+            )
         })
     }
 }
 
 pub struct Intersection<I: Iterator, J: Iterator> {
-    pub(crate) a: Peekable<I>,
+    pub(crate) a: I,
```

Comment:
Ah, so the optimization is that one side does not have to be wrapped in peekable. Nice.